Sign language Fixes & Spellcasting, just for fun

SylvetteSylph opened the pull request:
Fixes #72165
First PR, first-time coder, this might be a doozy but _hopefully_ everything is fine.

## About The Pull Request

This fixes signers being able to speak with sign language while cuffed or emotemute, bringing the feature back to how it was initially intended.

This also fixes signers only being able to sign spells based on their ability to speak. Before this PR, signers could cast spells with sign language, but it was dependent on their ability to speak with their mouth, allowing them to sign spells even if they would not be able to. Instead, it has been changed to work so that one can sign spells only when both of their hands are completely empty.
## Why It's Good For The Game

First of all, signers should probably not be able to speak when cuffed or emotemute, I'm not sure when this broke but somewhere along the lines it did and this is simply bringing it back to how it was supposed to be.

Second of all, spells were created long before sign langauge (I believe, don't quote me on that), but sign language is just another language and still can communicate the same. Furthermore, signing spells instead of speaking them is overall more difficult to work with given the fact that it won't work if you are holding items, restrained, have a limb disabled or amputated, or are emotemuted. The one benefit that sign language provides is being able to sign when otherwise mute, but this only really applies in select cases. I feel the downsides definitely outweigh the upsides in this case, and it will be more of a gimmick so that mute players can still play antags like heretic.
## Changelog
:cl: Sylvette
fix: Fixed sign language so you can't sign when cuffed or emotemute as initially intended
balance: Allows signers to cast spells using sign language as long as both hands are free
/:cl:

Especially thanks to Zonespace and Mal for coding assistance, and Inithis for necessary moral support

SylvetteSylph committed:
Sign language fixes.

First PR, this is gonna be a doozy I feel.

First off: Signers are not supposed to be able to speak when cuffed or emote-mute. This PR fixes that, and makes it work as initially intended.

Secondly: Spellcasters could only sign spells based on their ability to speak, not their ability to sign. This does not make sense, so it has been changed to work whenever you are capable of signing. This does provide the one bonus that you are able to do spells when normally mute (such as under mute toxin) but otherwise is much more difficult to work with, as both of your hands must be completely empty to cast.

by gods i hope this PRs to my own fork and not TG directly

SylvetteSylph committed:
you bastard, how did you get there

fixes the space indents lol

SylvetteSylph committed:
reducing code bloat in sign_language.dm

realized i could just use >=1 in this circumstance instead of putting everything manually and couldn't go to sleep until i fucking fixed and tested it

Zonespace27 commented on code/datums/components/sign_language.dm:
```diff
@@ -196,6 +197,11 @@
 
+/// Spellcasting with sign language
+/datum/component/sign_language/proc/can_cast_spell()
+	if(check_signables_state() >= 1) // Cannot cast if anything but SIGN_OKAY
```
```suggestion
	if(check_signables_state() >= SIGN_HANDS_FULL) // Cannot cast if anything but SIGN_OKAY
```

SylvetteSylph replied:
I'm not _100%_ sure about making it possible to cast if you have one hand open. Part of the balance was making it more difficult to use by requiring your hands be free, but considering you can speak spells fine while gibbering, it does sort of fit I suppose.

Zonespace27 replied:
do vinylspider's review for this line instead, it works better

Zonespace27 commented on code/modules/spells/spell.dm:
```diff
@@ -373,10 +373,18 @@
-			to_chat(owner, span_warning("You can't get the words out to invoke [src]!"))
-		return FALSE
+	if((invocation_type == INVOCATION_WHISPER || invocation_type == INVOCATION_SHOUT))
+		if(!HAS_TRAIT(living_owner,TRAIT_SIGN_LANG) && !living_owner.can_speak())
```
```suggestion
		if(!HAS_TRAIT(living_owner, TRAIT_SIGN_LANG) && !living_owner.can_speak())
```

Zonespace27 commented on code/modules/spells/spell.dm:
```diff
@@ -373,10 +373,18 @@
+			return FALSE
+		// Handle sign language
+		else if(HAS_TRAIT(living_owner, TRAIT_SIGN_LANG))
+			var/sigreturn = SEND_SIGNAL(living_owner, COMSIG_CARBON_TRY_SIGN_SPELL)
+			if(sigreturn & COMSIG_CANCEL_SIGN_SPELL)
```
```suggestion
			if(SEND_SIGNAL(living_owner, COMSIG_CARBON_TRY_SIGN_SPELL) & COMSIG_CANCEL_SIGN_SPELL)
```

Zonespace27 commented on code/modules/spells/spell.dm:
```diff
@@ -373,10 +373,18 @@
+			return FALSE
+		// Handle sign language
+		else if(HAS_TRAIT(living_owner, TRAIT_SIGN_LANG))
+			var/sigreturn = SEND_SIGNAL(living_owner, COMSIG_CARBON_TRY_SIGN_SPELL)
```
in the signal documentation, you list the first argument as `mob/user`, which is incorrect, as you never pass that in. If you are talking about the `datum/source` that is passed in with every signal, that shouldn't be included in the argument documentation in `signals_spell.dm`, as it is implied

SylvetteSylph replied:
Honestly speaking, I still don't know how to document signals properly, I was going off of what I was told would work for that. Would I just say it's going to datum/components/sign_language?

SylvetteSylph committed:
forgor space

space in if(!HAS_TRAIT(living_owner, TRAIT_SIGN_LANG) at 377

Co-authored-by: Zonespace <41448081+Zonespace27@users.noreply.github.com>

SylvetteSylph committed:
Simplifies signal

thank you zonespace bestie

Co-authored-by: Zonespace <41448081+Zonespace27@users.noreply.github.com>

SylvetteSylph committed:
Fixes signal documentation

Co-authored-by: Bloop <vinylspiders@gmail.com>

MrMelbert reviewed (changes requested):
Please check that this PR falls in line with the existing outline for sign language spell-casting mentioned here #72165

SylvetteSylph committed:
add SIGNAL_HANDLER

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

SylvetteSylph committed:
check_signables_state != SIGN_OKAY

Co-authored-by: Bloop <vinylspiders@gmail.com>

SylvetteSylph commented on code/__DEFINES/dcs/signals/signals_spell.dm:
```diff
@@ -14,6 +14,11 @@
 	/// Return from before cast signals to prevent the spell from going on cooldown before aftercast.
 	#define SPELL_NO_IMMEDIATE_COOLDOWN (1 << 2)
 
+/// Sent to an mob when a [/datum/action/cooldown/spell] calls try_invoke().
+#define COMSIG_CARBON_TRY_SIGN_SPELL "try_invoke_spell"
+	/// The spell gets canceled
+	#define COMSIG_CANCEL_SIGN_SPELL (1<<1)
```
The issue with doing it this way is that if a mime has the signer trait (and say mute trait, so they can't just toggle signing off), then their mime spells are no longer based on whether or not they can emote, but whether or not they can sign.

SylvetteSylph committed:
Makes signals more general & also bends over backwards to allow signing mimes to keep emote spells

Changes the sign signal to a general COMSIG_MOB_TRY_INVOKE_SPELL signal, adds SPELL_INVOCATION_FAIL and SPELL_INVOCATION_SUCCESS, and fucks it all up so mimes aren't affected by this.

SylvetteSylph committed:
Mimes shouldn't be able to cast spoken spells with sign language in vow of silence.

Mimes signing breaks their vow of silence, so a mime currently under the vow of silence shouldn't be able to sign a spell for obvious reasons.

SylvetteSylph committed:
Moves the jank out of spell.dm

Moves the schenanigans to determine if a spell is INVOCATION_EMOTE & the feedback for sign language spells all back to sign_language.dm.

SylvetteSylph committed:
Merge branch 'master' into sign_fixes

SylvetteSylph committed:
Documents proc for can_cast_spell()

Thanks to vinylspiders for this.

SylvetteSylph committed:
Merge branch 'sign_fixes' of https://github.com/SylvetteSylph/tgstation into sign_fixes

SylvetteSylph committed:
Fixes runtimes, whoops

Signal was initially not in list, causing a runtime every time one switched sign language on and off. This should also remove the necessity for checking a user for sign language when this goes through, as this should stop sign_language.dm from checking the signal when one does not have TRAIT_SIGN_LANG hopefully maybe probably

Sets the first arg to source so the INVOCATION_EMOTE check doesn't runtime and cause the entire sign check to get skipped. Also happens to fix feedback saying you cannot sign to invoke yourself.

SylvetteSylph committed:
Merge branch 'tgstation:master' into sign_fixes

SylvetteSylph committed:
Merge branch 'tgstation:master' into sign_fixes

san7890 commented on code/modules/spells/spell.dm:
```diff
@@ -368,6 +368,12 @@
 		return FALSE
 
 	var/mob/living/living_owner = owner
+	var/sigreturn = SEND_SIGNAL(owner, COMSIG_MOB_TRY_INVOKE_SPELL, src, feedback)
+	if(sigreturn & SPELL_INVOCATION_FAIL)
+		return FALSE
+	if(sigreturn & SPELL_INVOCATION_SUCCESS)
+		return TRUE // Skips following checks.
```
we can just do this like this:

```suggestion
	switch(SEND_SIGNAL(owner, COMSIG_MOB_TRY_INVOKE_SPELL, src, feedback))
		if(SPELL_INVOCATION_FAIL)
			return FALSE
		if(SPELL_INVOCATION_SUCCESS)
			return TRUE // skips all of the following checks
```

it has approximately the same line-footprint but I believe that this is a bit more verbose and clean, as well as being 0.000001 cpu cycle faster since we don't have to do two `if()` checks (as well as not holding onto a proc-local arg for the entirity of the rest of the proc), just one switch for both potential cases (and any abnormalities are handled exactly the same way).

san7890 replied:
i just noticed that melbert was the one to feed you this code but i still think that for the purposes and wider-scope of this proc that it's still a relatively good solution. ignore this if you truly want to but i'm fond of this

SylvetteSylph committed:
trailing commas

Co-authored-by: san7890 <the@san7890.com>

SylvetteSylph committed:
Remove unnecessary else

Co-authored-by: san7890 <the@san7890.com>

SylvetteSylph committed:
Yeah I like it too.

changes var/sigreturn to switch()

Co-authored-by: san7890 <the@san7890.com>

SylvetteSylph committed:
Merge branch 'tgstation:master' into sign_fixes

AverageUser67 commented:
Mime heretics should "gesticulate wildly" instead of doing voicelines

SylvetteSylph commented:
> Mime heretics should "gesticulate wildly" instead of doing voicelines

Mimes can't use sign language.

JohnFulpWillard commented on code/datums/components/sign_language.dm:
```diff
@@ -196,6 +198,29 @@
+	SIGNAL_HANDLER
+	var/mob/living/carbon/carbon_parent = parent
+	if(spell.invocation_type == INVOCATION_EMOTE) // Mime spells are not cast with signs
+		return
```
This is not returning success or fail

SylvetteSylph replied:
Vinylspiders is correct. This isn't intended to return success or fail, as mime spells (or any emote invocation spells for that matter, though I'm fairly certain this exists only for mime spells) aren't meant to go through this check at all. Without this, it means that if a mute sign language player chooses to play mime, their spells will also be forced to go through the sign language check despite not being spoken.

MrMelbert commented on code/modules/spells/spell.dm:
```diff
@@ -368,6 +368,12 @@
 		return FALSE
 
 	var/mob/living/living_owner = owner
+	switch(SEND_SIGNAL(owner, COMSIG_MOB_TRY_INVOKE_SPELL, src, feedback))
+		if(SPELL_INVOCATION_FAIL)
+			return FALSE
+		if(SPELL_INVOCATION_SUCCESS)
+			return TRUE // skips all of the following checks
```
```suggestion
	var/invoke_sig_return = SEND_SIGNAL(owner, COMSIG_MOB_TRY_INVOKE_SPELL, src, feedback)
	if(invoke_sig_return & SPELL_INVOCATION_ALWAYS_SUCCEED)
		return TRUE // skips all of the following checks
	if(invoke_sig_return & SPELL_INVOCATION_FAIL)
		return FALSE
```
Switch is not good since it's a bitflag, multiple signals might return multiple results.

MrMelbert commented on code/__DEFINES/dcs/signals/signals_spell.dm:
```diff
@@ -14,6 +14,13 @@
+#define COMSIG_MOB_TRY_INVOKE_SPELL "try_invoke_spell"
+	/// The spell gets canceled
+	#define SPELL_INVOCATION_FAIL SPELL_CANCEL_CAST
+	/// The spell succeeds
+	#define SPELL_INVOCATION_SUCCESS (1 << 1)
```
```suggestion
	/// The spell always succeeds to invoke regardless of following checks
	#define SPELL_INVOCATION_ALWAYS_SUCCEED (1 << 1)
```

MrMelbert commented on code/datums/components/sign_language.dm:
```diff
@@ -196,6 +198,29 @@
+	SIGNAL_HANDLER
+	var/mob/living/carbon/carbon_parent = parent
+	if(spell.invocation_type == INVOCATION_EMOTE) // Mime spells are not cast with signs
+		return
```
```suggestion
		return NONE // Run normal checks
```

MrMelbert commented on code/datums/components/sign_language.dm:
```diff
@@ -196,6 +198,29 @@
+			to_chat(carbon_parent, span_warning("You can't sign the words to invoke [spell]!"))
+		return SPELL_INVOCATION_FAIL
+
+	return SPELL_INVOCATION_SUCCESS
```
```suggestion
	return SPELL_INVOCATION_ALWAYS_SUCCEED
```

MrMelbert reviewed (commented):
Sorry for letting this by the wayside

MrMelbert committed:
Assuming direct control
